### ISSUE TITLE
Architect: SAC trainer for continuous control (ROADMAP Milestone 6)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#136]
## Background

Thrust's algorithm portfolio is currently:
- **PPO** (on-policy, single-agent + multi-agent via \`JointMultiAgentTrainer\`)
- **DQN** (off-policy, single-agent discrete)
- **PSRO with α-rank** (multi-agent N-player meta-solver, via PR #123/#125)
- **NFSP (approximate, N-player)** (PR #123, multi-discrete after #132)

**Gap**: no continuous-control algorithm. ROADMAP.md Milestone 6 calls out **SAC** (Soft Actor-Critic) as one of the canonical adds, alongside A2C and behavioral cloning. SAC is the higher-impact addition because it (a) unlocks all the continuous-control RL benchmarks (LunarLanderContinuous, MuJoCo, Box2D) and (b) is the de facto SOTA off-policy method for continuous actions.

This is /architect work, not /builder, because:

1. **Action space generalization**. The current \`Environment\` trait uses \`SpaceType::Discrete(n)\` / \`SpaceType::Continuous { ... }\`. PPO's discrete action path is exercised; continuous actions are technically supported but no trainer has been built against them. The trait may need a small clarifying change (e.g., a \`ContinuousAction\` associated type, or a \`Vec<f32>\` convention).

2. **Replay buffer design**. SAC is off-policy and needs a replay buffer (not the rollout buffer PPO uses). The existing \`src/buffer/rollout/\` is on-policy-only. A new \`src/buffer/replay/\` module is needed. Decision: shared API with rollout buffer (generic over storage strategy), or completely separate trait surface?

3. **Twin Q-networks + target networks**. SAC uses 2 Q-critics + 2 target Q-networks + 1 stochastic actor (5 networks total per agent). The current \`MlpBurnPolicy\` is single-network. We need either a new policy type (\`SacPolicy\`) or a generic actor-critic factory.

4. **Continuous-action env**. We have no continuous-action env in-tree right now. The architect needs to decide: build a small in-tree continuous env (e.g., \`PendulumSwingUp\`, ~150 LOC), OR pull in \`gym-rs\` / similar for Atari/MuJoCo bindings, OR use a custom env from the bucket-brigade repo.

## Scope

Produce a written architectural proposal that answers:

1. **Action-space trait extension**. Does \`Environment\` need a new associated type, or does \`Vec<f32>\` per-step suffice with the existing trait? What's the minimal change that keeps PPO/DQN/PSRO/NFSP working bit-stably?

2. **Replay buffer surface**. New trait \`ReplayBuffer\` parallel to the rollout buffer? Or generic \`Buffer<S: Strategy>\` with strategies \`Rollout\` and \`Replay\`? Trade off composability vs surface area.

3. **SAC policy architecture**. New \`SacPolicy<B: Backend>\` type with 5 sub-modules? Or a factory function that produces 5 \`MlpBurnPolicy\` instances and threads them through a \`SacTrainer\`? Decide between encapsulation vs reuse.

4. **Hyperparameter set**. SAC has a lot of knobs (\`alpha\` entropy temperature, automatic-vs-fixed entropy, target update tau, batch size, replay buffer capacity, ...). What's the canonical set the trainer should expose? Reference: Haarnoja et al. 2018 \"Soft Actor-Critic\" (arXiv:1801.01290) and the SAC v2 paper (\"Soft Actor-Critic Algorithms and Applications\", arXiv:1812.05905).

5. **First env target**. Pick ONE of:
   - **In-tree PendulumSwingUp** (~150 LOC, no external deps; closed-form physics).
   - **External Atari/MuJoCo bindings** (substantial effort, third-party deps).
   - **A bucket-brigade continuous variant** (small effort, but couples SAC to multi-agent which is not its strength).

6. **Test plan**. What's the matching-pennies-equivalent smoke test for SAC? Pendulum is the natural choice (small state, fast to convergence, well-characterized). Define the smoke test bar (e.g., \"SAC on Pendulum reaches ≥ -200 mean reward over 100 episodes after 50k env steps\").

7. **Decomposition into Builder-tractable PRs**. Order of work + blast radius per PR. Target 3-5 PRs.

## Out of scope (for the proposal)

- Implementing anything. Builder issues for the implementation come after the proposal lands.
- A2C and behavioral cloning (separate Milestone 6 items; can be future architect issues).
- Distributed/multi-GPU SAC (Milestone 7).
- Hyperparameter sweep infrastructure (separate concern).

## Acceptance criteria

- [ ] Written proposal (as an issue comment) addressing items 1-7 above.
- [ ] Operator approves a path with a sign-off comment.
- [ ] One or more follow-up Builder-tractable issues filed implementing the chosen path.

## Source

ROADMAP.md Milestone 6 (Algorithm Expansion). First continuous-control algorithm; opens the door to a much wider env catalog and benchmark coverage.

\`loom:architect\` — needs design pass before any Builder can act.
\`tier:goal-advancing\` — meaningful capability expansion.

**Comment by rjwalters** [comment]
## Architectural Proposal — SAC trainer for continuous control

Operator approved decomposition. This proposal answers items 1–7 against the **current** codebase, which is further along than the issue text assumes: the `Environment` trait already carries an associated `Action` type (Strategy A, documented in `src/env/mod.rs`), an in-tree continuous env (`src/env/games/continuous_lqr.rs`) already exists as a `type Action = Vec<f32>` proof, and a discrete-only replay buffer already lives at `src/buffer/replay/`. The design below leans into that existing surface rather than re-litigating it.

### 1. Action-space trait extension — NO trait change needed

The hard work is already done. `Environment` (`src/env/mod.rs:66`) exposes `type Action;`; discrete envs set `type Action = i64`, and `ContinuousLqr` already sets `type Action = Vec<f32>` with `SpaceType::Box` action space. The module doc explicitly records this as "Strategy A" chosen precisely to unblock SAC/TD3/DDPG.

**Decision:** No change to the `Environment` trait, `SpaceType`, `StepResult`, or `SpaceInfo`. SAC's trainer bounds its env type parameter as `E: Environment<Action = Vec<f32>>`, exactly mirroring how discrete trainers will bound `E: Environment<Action = i64>`. PPO/DQN/PSRO/NFSP are untouched and stay bit-stable because their generic instantiations are unchanged. The only convention SAC adds is: a Box action env reports `action_space().shape == vec![action_dim]` and accepts a `Vec<f32>` of that length, with the env responsible for clamping to its valid range (as `ContinuousLqr` already does). This is the smallest viable change: zero lines in the trait.

### 2. Replay buffer surface — extend the replay module with a continuous-action buffer; do NOT unify with rollout

The existing `ReplayBuffer` (`src/buffer/replay/storage.rs`) hard-codes `actions: Vec<i64>` and `push(.., action: i64, ..)`. SAC needs `Vec<f32>` actions of width `action_dim`. The rollout buffer (`src/buffer/rollout/`) is a fundamentally different beast (GAE, on-policy, value/logprob/advantage columns); a generic `Buffer<S: Strategy>` would force the union of both column sets behind a strategy enum and bloat surface area for no caller benefit — nothing in the repo wants to swap rollout/replay polymorphically.

**Decision:** Keep rollout and replay as separate concrete types (status quo). Add a sibling **`ContinuousReplayBuffer`** in `src/buffer/replay/` with the same ring/FIFO semantics and the same flat-`Vec<f32>` WASM-compatible storage as `ReplayBuffer`, but with `actions: Vec<f32>` of length `capacity * action_dim` and a `push(&[f32] obs, &[f32] action, f32 reward, &[f32] next_obs, bool done)` signature. Provide a parallel `ContinuousReplayBatch` + `sample()` + `to_burn_tensors()` mirroring `src/buffer/replay/sampling.rs`. Rationale: maximal reuse of the proven ring-buffer/sampling design, zero risk to the discrete DQN path, no new trait surface. (A future refactor could factor the ring mechanics into a shared private helper, but that is out of scope and not required for SAC.)

### 3. SAC policy architecture — dedicated `SacActor` + reuse `QNetworkBurn`-style critics; assembled by the trainer

SAC needs: 1 stochastic Gaussian actor (mean + log-std heads, tanh-squashed) and 2 Q-critics + 2 target critics. The critics are continuous-input `Q(s,a)` networks: they take `concat(obs, action)` and output a scalar — structurally a plain MLP with a single head, very close to `QNetworkBurn` but with input width `obs_dim + action_dim` and output width `1` (vs `QNetworkBurn`'s `obs_dim -> n_actions`).

**Decision:** Two new small policy modules in `src/policy/`:
- **`SacActor<B>`** (`src/policy/sac_actor.rs`): shared trunk + `mean_head` + `log_std_head` (both `hidden -> action_dim`). Exposes `sample(obs) -> (action, log_prob)` using the reparameterized tanh-squashed-Gaussian with the standard `log_prob -= sum(log(1 - tanh(u)^2 + eps))` correction, plus a deterministic `mean_action(obs)` for eval. Reuses `linear_with_init` from `mlp.rs`.
- **`ContinuousQNetwork<B>`** (`src/policy/continuous_q.rs`): `concat(obs,action) -> trunk -> scalar`. Reuses `linear_with_init`. The trainer owns FOUR instances (q1, q2, q1_target, q2_target) plus a `copy_params_from` / soft-blend helper analogous to `QNetworkBurn`'s record-based target sync.

A monolithic `SacPolicy` bundling all 5 nets as one `Module` is rejected: Burn's optimizer is move-through and consumes a module by value per step; SAC needs THREE independent optimizers (actor, critics, and alpha) stepped separately, so the five nets must be owned as separate module fields by the trainer, not one fused module. This is the factory-of-small-modules approach (option B in the issue), and it composes with the existing `BurnOptimizer`/`BackendOptimizer` wrapper unchanged.

### 4. Hyperparameter set — `SacConfig`

Mirrors `DQNConfig`'s builder + `validate()` style. Defaults track Haarnoja et al. 2018 v2 (arXiv:1812.05905) on Pendulum-scale tasks:

| Knob | Default | Notes |
|---|---|---|
| `actor_lr` | `3e-4` | Adam |
| `critic_lr` | `3e-4` | Adam |
| `alpha_lr` | `3e-4` | Adam, only used when auto-tuning |
| `gamma` | `0.99` | discount |
| `tau` | `0.005` | Polyak soft target update (always soft in SAC) |
| `batch_size` | `256` | SAC standard |
| `buffer_capacity` | `1_000_000` | (smoke tests override to ~50k) |
| `min_buffer_size` | `1_000` | warmup before first gradient step |
| `learning_starts` | `1_000` | random-action exploration steps before policy actions |
| `gradient_steps_per_env_step` | `1` | UTD ratio |
| `hidden_dim` | `256` | actor + critics |
| `num_hidden_layers` | `2` | |
| `auto_alpha` | `true` | automatic entropy tuning (v2 paper) |
| `init_alpha` | `0.2` | initial/fixed entropy temperature |
| `target_entropy` | `None` → `-action_dim` | conventional heuristic when auto-tuning |
| `max_grad_norm` | `None` | optional clip (SAC usually unclipped) |
| `seed` | `0` | bit-exact reproducibility, like `PsroConfig::seed` |

`validate()` checks: positive LRs, `0<tau<=1`, `0<=gamma<=1`, `batch_size>0`, `buffer_capacity>=batch_size`, `min_buffer_size<=buffer_capacity`, `init_alpha>0`.

### 5. First env target — in-tree `PendulumSwingUp`

Strongly prefer the issue's option A. No external deps, ~150 LOC, closed-form physics, the canonical SAC smoke benchmark, and well-characterized convergence bar. Implement as `src/env/games/pendulum.rs` with `type Action = Vec<f32>` (length 1, torque in `[-2, 2]`), `type State` snapshot (deterministic, no RNG except initial-angle reset which we seed), 3-dim observation `[cos θ, sin θ, θ̇]`, reward `-(θ² + 0.1·θ̇² + 0.001·u²)`, 200-step episodes. `ContinuousLqr` stays as the trait existence-proof; Pendulum becomes the real training target. Atari/MuJoCo bindings and bucket-brigade-continuous are rejected (deps / multi-agent coupling).

### 6. Test plan — Pendulum smoke test (matching-pennies analogue)

A `tests/test_sac_pendulum.rs` integration test, gated on the `training` feature like the PSRO/NFSP smoke tests. Bar:

> SAC on `PendulumSwingUp`, seeded, reaches **mean episode reward ≥ -200 over the final 20 evaluation episodes after 30k env steps**.

Pendulum's near-optimal return is roughly `-150`; a uniformly-bad policy sits around `-1200..-1600`, so `-200` cleanly separates "learned" from "did not learn" and is reproducible in CI at 30k steps with the default config (capacity overridden to 50k for the test). Additional fast unit tests (no convergence): actor `sample()` returns actions in `(-1,1)·scale` with finite log-probs; `log_prob` tanh-correction shape; target soft-update moves params toward online; `ContinuousReplayBuffer` round-trip + ring eviction (mirroring the discrete buffer tests); `SacConfig::validate()` rejection cases.

### 7. Decomposition into Builder-tractable PRs

Five PRs, ordered. Dependency DAG: **A → {B, C, D} → E**.

- **PR A — `ContinuousReplayBuffer` + sampling** (`src/buffer/replay/`). Blast radius: additive, new module files; no existing buffer touched. No deps. *Ready to build.*
- **PR B — `PendulumSwingUp` env** (`src/env/games/pendulum.rs` + re-export in `src/env/mod.rs`/`games/mod.rs`). Blast radius: additive new env. No deps. *Ready to build.*
- **PR C — `SacActor` policy module** (`src/policy/sac_actor.rs`). Blast radius: additive new policy. No deps. *Ready to build.*
- **PR D — `ContinuousQNetwork` critic + target-sync helper** (`src/policy/continuous_q.rs`). Blast radius: additive new policy. No deps. *Ready to build.* (B/C/D are independent of each other; only the trainer needs all of them.)
- **PR E — `SacTrainer` + `SacConfig` + Pendulum smoke test** (`src/train/sac/{mod,config,trainer}.rs`, `tests/test_sac_pendulum.rs`). Wires actor + 4 critics + 3 optimizers + `ContinuousReplayBuffer` into the env loop; auto-alpha; soft target updates; the item-6 smoke test. Blast radius: large but self-contained new trainer module. **Depends on A, B, C, D.**

This keeps each of A–D under a few hundred LOC and individually mergeable/testable; E is the only integration-heavy PR and lands last once its four inputs are on `main`.

---
Child implementation issues filed per this decomposition; numbers + DAG posted in a follow-up comment. #136 stays open as the tracking issue.

**Comment by rjwalters** [comment]
## Child implementation issues filed

Decomposition from item 7 is now on GitHub. Dependency DAG: **{A, B, C, D} → E**.

| PR | Issue | Title | Label | Depends on |
|---|---|---|---|---|
| A | #138 | ContinuousReplayBuffer | `loom:issue` (ready) | — |
| B | #139 | PendulumSwingUp env | `loom:issue` (ready) | — |
| C | #140 | SacActor (tanh-Gaussian policy) | `loom:issue` (ready) | — |
| D | #141 | ContinuousQNetwork critic + target sync | `loom:issue` (ready) | — |
| E | #142 | SacTrainer + SacConfig + Pendulum smoke test | `loom:blocked` | #138, #139, #140, #141 |

```
#138 ─┐
#139 ─┤
#140 ─┼──> #142 (SacTrainer + smoke test)
#141 ─┘
```

**Immediately buildable (no blockers):** #138, #139, #140, #141 — all labeled `loom:issue`, fully independent of each other.

**Blocked until its four inputs merge:** #142 — labeled `loom:blocked`; the sweep skips it until A–D land on `main`, at which point it should be flipped to `loom:issue`.

#136 remains open as the tracking issue (keeps `loom:architect`).

**Comment by rjwalters** [comment]
## SAC epic complete ✅

All decomposed child issues have been implemented, reviewed, and merged to `main`:

| Issue | PR | Component |
|-------|-----|-----------|
| #138 | #143 | `ContinuousReplayBuffer` (off-policy continuous-action storage) |
| #139 | #144 | `PendulumSwingUp` continuous-control env |
| #140 | #145 | `SacActor` (tanh-squashed Gaussian policy) |
| #141 | #146 | `ContinuousQNetwork` (twin-critic + Polyak target sync) |
| #142 | #147 | `SacTrainer` + `SacConfig` + Pendulum smoke test |

Soft Actor-Critic (Haarnoja et al. 2018 v2) now ships end-to-end: clipped double-Q critics, reparameterized actor loss, automatic entropy-temperature tuning, three independent optimizers, and deterministic seeding throughout. The convergence smoke test reaches the ≥ -200 mean-reward bar on PendulumSwingUp after 30k env steps (verified in release, ~273s).

A latent flaky test from #146 (`soft_update_moves_target_toward_online` asserted an output-space property Polyak updates don't guarantee) was also fixed in #147 — now asserts the parameter-space invariant with seeded determinism.

Closing this tracking issue; the proposal (items 1-7) and DAG are in the comments above. Follow-on Milestone 6 items (A2C, behavioral cloning) remain as future `/architect` work per the original scope.
